File: generator/stdlib_ml/stdlib.js
// JS implementations for OCaml's Pervasives library
// See stdlib.mli for documentation

// Infix operators are converted to equivalent JS in
//   js_of_ast.ml#js_of_path_longident

//----------------------------------------------------------------------------
// Js_of_ocaml bis Generator Requirements

// type option 'a = None | Some of 'a
var None = function() {
   return { /*type: "option",*/ tag: "None" };
};

var Some = function(value) {
   return { /*type: "option",*/ tag: "Some", value: value };
};

// type list 'a = [] | :: of 'a * list 'a
var mk_nil = function() {
   return { /*type: "list",*/ tag: "[]" };
};

var mk_cons = function(head, tail) {
   return { /*type: "list",*/ tag: "::", head: head, tail: tail };
};

//----------------------------------------------------------------------------
// Exceptions

// val raise : exn -> 'a
var raise = function(x) { throw x; };
var failwith = function(str) { throw Error(str); };

//----------------------------------------------------------------------------
// Boolean operations

// val not : bool -> bool
var not = function(x) { return !x; };

// val ( && ) : bool -> bool -> bool
// val ( || ) : bool -> bool -> bool
// Operators mapped directly to JS equivalent

//----------------------------------------------------------------------------
// Deugging

// val __LOC__ : string
var __LOC__ = "___LOC___"

//----------------------------------------------------------------------------
// Integer Arithmetic

// val ( + ) : int -> int -> int
// val ( - ) : int -> int -> int
// val ( * ) : int -> int -> int
// val ( / ) : int -> int -> int
// Operators mapped directly to JS equivalent

//----------------------------------------------------------------------------
// Floating-point Arithmetic

// val ( +. ) : float -> float -> float
// val ( -. ) : float -> float -> float
// val ( *. ) : float -> float -> float
// val ( /. ) : float -> float -> float
// Operators mapped to JS equivalent by removing .

//----------------------------------------------------------------------------
//----------------------------------------------------------------------------

// Pervasives-incompatible functions

// Infix operators are converted to equivalent JS in
//   js_of_ast.ml#js_of_path_longident

//----------------------------------------------------------------------------
// Comparisons

// val ( = ) : float -> float -> bool
// Mapped to ==

// val ( < ) : float -> float -> bool
// val ( > ) : float -> float -> bool
// val ( <= ) : float -> float -> bool
// val ( >= ) : float -> float -> bool
// Operators mapped directly to JS equivalent

/*
// val compare : 'a -> 'a -> int
// val min : float -> float -> float
// val max : float -> float -> float
*/

// val ( === ) : 'a -> 'a -> bool

/*
// val float_compare : float -> float -> int
*/

// val int_eq : int -> int -> bool
var int_eq = function(x, y) { return x === y; };

// val int_lt : int -> int -> bool
var int_lt = function(x, y) { return x < y; };

// val int_gt : int -> int -> bool
var int_gt = function(x, y) { return x > y; };

// val int_le : int -> int -> bool
var int_le = function(x, y) { return x <= y; };

// val int_ge : int -> int -> bool
var int_ge = function(x, y) { return x >= y; };

// val int_compare : int -> int -> int
var int_compare = function(x, y) { return x - y; };

// val bool_eq : bool -> bool -> bool
var bool_eq = function(x, y) { return x === y; };

// val nat_eq : int -> int -> bool (* nat_eq x y = int_eq x y  *)
var nat_eq = function(x, y) { return x === y; };

// val string_eq : string -> string -> bool (* === *)
var string_eq = function(x, y) {
  if (typeof(x) != "string" || typeof(y) != "string")
    throw "string_eq invalid arguments";
  return x === y;
};

// val string_lt : string -> string -> bool
var string_lt = function(x, y) {
  if (typeof(x) != "string" || typeof(y) != "string")
    throw "string_lt invalid arguments";
  return x < y;
};

// val string_compare : string -> string -> int
var string_compare = function(x, y) {
  return x.localeCompare(y);
};

//----------------------------------------------------------------------------
// Floating-point arithmetic

/* Functions renamed:
// val fmod : float -> float -> float (*  mod_float, implemented as % operator in JS *)
// val float_neg : float -> float          (* ~-. *)
// val float_exp : float -> float -> float (* exp *)
*/

// Alan: Ideally we would add these to the spec, but for the moment conversion
// to a string is doing a foo+"", and conversion to an int is doing +foo
// val int_of_number : float -> int (* will be removed, since only used by substring *)
var int_of_number = function(x) { return x; };

// val number_of_int : int -> float  (* = fun x -> float_of_int x *)
var number_of_int = function(x) { return x; };

//----------------------------------------------------------------------------
// String operations

// val string_concat : string -> string -> string (* + *)

// val strappend : string -> string -> string
var strappend = function(x, y) {
  if (typeof(x) != "string" || typeof(y) != "string")
    throw "strappend invalid arguments";
  return x + y;
};

// val strlength : string -> int
var strlength = function(x) {
  if (typeof(x) != "string")
    throw "strlength invalid arguments";
  return x.length;
};

// val substring : int -> int -> string -> string
var substring = function(n, m, s) {
  if (typeof(s) != "string")
    throw "strlength invalid arguments";
  return s.slice(n, n+m);
};

File: generator/stdlib_ml/stdlib.ml
(** OCaml Compatibility Wrapper for the Js_of_ocaml bis Standard Library

This implementation of the library permits programs designed to be compiled
with the limited standard library of Js_of_ocaml bis to be compiled against the
OCaml standard library for native/bytecode targets.

This file is to be compiled with the standard OCaml compiler with the
-nopervasives flag enabled. It should be linked with the OCaml standard
library into the end program file.
*)
let raise = Pervasives.raise;;

(**{6 Boolean operations }*)
(** Note: Both OCaml and JS implement lazy evaluation for boolean operators. *)
let not = Pervasives.not;;
external ( && ) : bool -> bool -> bool = "%sequand";;
external ( || ) : bool -> bool -> bool = "%sequor";;

(**{6 Debugging }*)
external __LOC__ : string = "%loc_LOC"

(**{6 Integer arithmetic }*)
let ( + ) = Pervasives.( + );;
let ( - ) = Pervasives.( - );;
let ( * ) = Pervasives.( * );;
let ( / ) = Pervasives.( / );;

(**{6 Floating-point arithmetic }*)
let ( +. ) = Pervasives.( +. );;
let ( -. ) = Pervasives.( -. );;
let ( *. ) = Pervasives.( *. );;
let ( /. ) = Pervasives.( /. );;

(*
let ( ** ) = Pervasives.( ** );;
let atan = Pervasives.atan;;
let exp = Pervasives.exp;;
let log = Pervasives.log;;
let mod_float = Pervasives.mod_float;;
let float_of_int = Pervasives.float_of_int;;
let infinity = Pervasives.infinity;;
let neg_infinity = Pervasives.neg_infinity;;
let nan = Pervasives.nan;;
let max_float = Pervasives.max_float;;
let min_float = Pervasives.min_float;;
*)

(**{6 String operations }*)
(*
let (^) = Pervasives.(^);;
*)

(**{6 Character operations }*)
(*
let int_of_char = Pervasives.int_of_char;;
*)

(**{6 String conversion functions }*)
(*
let string_of_int = Pervasives.string_of_int;;
let string_of_float = Pervasives.string_of_float;;
let float_of_string = Pervasives.float_of_string;;
*)

(**{6 Input/output }*)
(*
let print_endline = Pervasives.print_endline;;
let prerr_string = Pervasives.prerr_string;;
let prerr_endline = Pervasives.prerr_endline;;
let prerr_newline = Pervasives.prerr_newline;;
*)

(**{6 References }*)
(** for future use for the global heap
let ref = Pervasives.ref;;
let (:=) = Pervasives.(:=);;
let (!) = Pervasives.(!);;
*)

(**{5 Pervasives-incompatible Definitions }

Functions in this section either deviate from the OCaml Pervasives type
signature, or are additional functions to fill in holes left by making a
polymorphic function not monomorphic.
*)

(**{6 Comparisons }*)
(** The standard comparison operators have been restricted to floating-point
operations only. *)
let ( = ) = Pervasives.( = );;
let ( < ) = Pervasives.( < );;
let ( > ) = Pervasives.( > );;
let ( <= ) = Pervasives.( <= );;
let ( >= ) = Pervasives.( >= );;

(*
let compare = Pervasives.compare;;
let min = Pervasives.min;;
let max = Pervasives.max;;
*)

let ( === ) = Pervasives.( = );;

(*
let float_compare = Pervasives.compare;;
*)
let int_eq = Pervasives.(=);;
let int_lt = Pervasives.(<);;
let int_gt = Pervasives.(>);;
let int_le = Pervasives.(<=);;
let int_ge = Pervasives.(>=);;
let int_compare = Pervasives.compare;;
let bool_eq = Pervasives.(=);;
let nat_eq = Pervasives.(=);;

let string_eq = Pervasives.(=);;
let string_lt = Pervasives.(<);;
let string_compare = Pervasives.compare;;

(**{6 Integer arithmetic }*)
(*let int_abs = Pervasives.abs;;*)

(**{6 Floating-point arithmetic }*)
(*
let fmod = Pervasives.mod_float;;
let float_neg = Pervasives.(~-.);;
let float_exp = Pervasives.exp;;
*)

(* Alan: Ideally we would add these to the spec, but for the moment conversion
   to a string is doing a foo+"", and conversion to an int is doing +foo *)
let int_of_number = Pervasives.int_of_float;;
let number_of_int = Pervasives.float_of_int;;

(**{6 String Operations }*)
(** String concatenation renamed from ^ *)
let strappend = Pervasives.(^);;

(**{5 String library }*)
(** Operations here are present in the OCaml standard library's String module
However cannot be used directly due to a potential module name collision when
linking to the standard library. (TODO: Work out if this can be avoided) *)

(*
This has the wrong typesignature for String.concat!
val string_concat : string -> string -> string (* + *)
*)

let strlength = String.length;;
let substring n m s = String.sub s n m;;

let console_int = Pervasives.print_int;;
